proposals/NNNN-init-accessors.md

```diff
@@ -0,0 +1,362 @@
+
+Swift applies [definite initialization analysis](https://en.wikipedia.org/wiki/Definite_assignment_analysis) to stored properties, stored local variables, and variables with property wrappers. Definite initialization ensures that memory is initialized on all paths before it is accessed. A common pattern in Swift code is to use one property as backing storage for one or more computed properties, and abstractions like [property wrappers](https://github.com/apple/swift-evolution/blob/main/proposals/0258-property-wrappers.md) and now [attached macros](https://github.com/apple/swift-evolution/blob/main/proposals/0389-attached-macros.md) help facilitate this pattern. Under this pattern, the backing storage is an implementation detail, and most code works with the computed property, including initializers.
+
+Property wrappers support bespoke definite initialization that allows initializing the backing property wrapper storage via the computed property, always re-writing initialization-via-wrapped-property in the form `self.value = value` to initialization of the backing storage in the form of `_value = Wrapper(wrappedValue: value)`:
+
+```swift
+@propertyWrapper
+struct Wrapper<T> {
+  var wrappedValue: T
+}
+
+struct S {
+  @Wrapper var value: Int
+
+  init(value: Int) {
+    self.value = value  // Re-written to self._x = Wrapper(wrappedValue: value)
+  }
+
+  init(other: Int) {
+    self._value = Wrapper(wrappedValue: other) // Okay, initializes storage '_x' directly
+  }
+}
+```
+
+The ad-hoc nature of property wrapper initializers mixed with an exact definite initialization pattern prevent property wrappers with additional arguments from being initialized out-of-line. Furthermore, property-wrapper-like macros cannot achieve the same initializer usability, because any backing storage variables added must be initialized directly instead of supporting initialization through computed properties. For example, the proposed [`@Observable` macro](https://github.com/apple/swift-evolution/blob/main/proposals/0395-observability.md) applies a property-wrapper-like transform that turns stored properties into computed properties backed by the observation APIs, but it provides no way to write an initializer using the original property names like the programmer expects:
+
+```swift
+@Observable
+struct Proposal {
+  var title: String
+  var text: String
+
+  init(title: String, text: String) {
+    self.title = title // error: 'self' used before all stored properties are initialized
+    self.text = text // error: 'self' used before all stored properties are initialized
+  } // error: Return from initializer without initializing all stored properties
+}
+```
+
+## Proposed solution
+
+This proposal adds _`init` accessors_ to opt computed properties on types into virtual initialization that subsumes initialization of a set of zero or more specified stored properties, which allows assigning to computed properties in the body of a type's initializer:
```

Comment:
>Is 'virtual' meant to be precise about the fact that the initialization is not (necessarily) tied to memory initialization of some storage location?

Yes, exactly. But I agree it's probably an unnecessary bit of jargon, and I can replace the phrase with something more direct.

Reply:
Yeah I think that's perhaps an interesting aside to include somewhere in this document but IMO we don't need to harp on the fact that this is 'virtual' initialization since part of the _point_ of this proposal is to lift DI into a language concept _separate_ from just the stored properties that need to be initialized for memory safety reasons. I think we should instead orient towards the fact that computed properties with `init` accessors are, for DI purposes, treated the _same_ as stored properties rather than using language to set them apart throughout.

Reply:
I removed the phrase "virtual initialization" throughout the document.